src/shared/mkfs-util.c

```diff
@@ -65,31 +65,27 @@ int make_filesystem(
         if (r < 0)
                 return r;
         if (r == 0) {
-                char suuid[ID128_UUID_STRING_MAX];
-
                 /* Child */
-                id128_to_uuid_string(uuid, suuid);
-
                 if (streq(fstype, "ext4"))
                         (void) execlp(mkfs, mkfs,
-                               "-L", label,
-                               "-U", suuid,
-                               "-I", "256",
-                               "-O", "has_journal",
-                               "-m", "0",
-                               "-E", discard ? "lazy_itable_init=1,discard" : "lazy_itable_init=1,nodiscard",
-                               node, NULL);
+                                      "-L", label,
+                                      "-U", ID128_TO_UUID_STRING(uuid),
+                                      "-I", "256",
+                                      "-O", "has_journal",
+                                      "-m", "0",
+                                      "-E", discard ? "lazy_itable_init=1,discard" : "lazy_itable_init=1,nodiscard",
+                                      node, NULL);
 
                 else if (streq(fstype, "btrfs")) {
                         if (discard)
-                                (void) execlp(mkfs, mkfs, "-L", label, "-U", suuid, node, NULL);
+                                (void) execlp(mkfs, mkfs, "-L", label, "-U", ID128_TO_UUID_STRING(uuid), node, NULL);
                         else
-                                (void) execlp(mkfs, mkfs, "-L", label, "-U", suuid, "--nodiscard", node, NULL);
+                                (void) execlp(mkfs, mkfs, "-L", label, "-U", ID128_TO_UUID_STRING(uuid), "--nodiscard", node, NULL);
```

Comment:
This will work nicely if the compiler knows that ID128_TO_UUID_STRING… is pure. If we don't have annotations, we should add them.

Reply:
Hmm, so we don't. Should we mark `sd_id128_to_string` as `__pure__`?

Reply:
Note that in this case each possible codepath through the function will only call sd_id128_to_string() only once, hence `_pure_` won't give you much benefit.

I am not sure it's really a great idea to put `_pure_` on public, exported functions, given that they leak implementation details in a way. I'd be careful with that, i.e. caching of stuff becomes hard if you encode publicly too much about your behavior inside.